translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,301 @@
+	"github.com/golang/protobuf/ptypes/wrappers"
+	otlpcommon "github.com/open-telemetry/opentelemetry-proto/gen/go/common/v1"
+	otlpmetrics "github.com/open-telemetry/opentelemetry-proto/gen/go/metrics/v1"
+	otlpresource "github.com/open-telemetry/opentelemetry-proto/gen/go/resource/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+)
+
+const (
+	invalidMetricDescriptorType = ocmetrics.MetricDescriptor_Type(-1)
+)
+
+// ResourceMetricsToMetricsData converts metrics from OTLP to internal (OpenCensus) format
+func ResourceMetricsToMetricsData(resourceMetrics *otlpmetrics.ResourceMetrics) consumerdata.MetricsData {
+
+	node, resource := resourceToOC(resourceMetrics.Resource)
+	md := consumerdata.MetricsData{
+		Node:     node,
+		Resource: resource,
+	}
+
+	if len(resourceMetrics.Metrics) == 0 {
+		return md
+	}
+
+	metrics := make([]*ocmetrics.Metric, 0, len(resourceMetrics.Metrics))
+	for _, metric := range resourceMetrics.Metrics {
+		metrics = append(metrics, metricToOC(metric))
+	}
+
+	md.Metrics = metrics
+	return md
+}
+
+func resourceToOC(resource *otlpresource.Resource) (*occommon.Node, *ocresource.Resource) {
+	// TODO extract from trace translator
+	panic("Not implemented")
+}
+
+func metricToOC(metric *otlpmetrics.Metric) *ocmetrics.Metric {
+	return &ocmetrics.Metric{
+		MetricDescriptor: descriptorToOC(metric.MetricDescriptor),
+		Timeseries:       dataPointsToTimeseries(metric),
+		Resource:         nil,
+	}
+}
+
+func descriptorToOC(descriptor *otlpmetrics.MetricDescriptor) *ocmetrics.MetricDescriptor {
+	return &ocmetrics.MetricDescriptor{
+		Name:        descriptor.Name,
+		Description: descriptor.Description,
+		Unit:        descriptor.Unit,
+		Type:        desctiptorTypeToOC(descriptor.Type),
+	}
+}
+
+func desctiptorTypeToOC(t otlpmetrics.MetricDescriptor_Type) ocmetrics.MetricDescriptor_Type {
+	switch t {
+	case otlpmetrics.MetricDescriptor_UNSPECIFIED:
+		return ocmetrics.MetricDescriptor_UNSPECIFIED
+	case otlpmetrics.MetricDescriptor_GAUGE_INT64:
+		return ocmetrics.MetricDescriptor_GAUGE_INT64
+	case otlpmetrics.MetricDescriptor_GAUGE_DOUBLE:
+		return ocmetrics.MetricDescriptor_GAUGE_DOUBLE
+	case otlpmetrics.MetricDescriptor_GAUGE_HISTOGRAM:
+		return ocmetrics.MetricDescriptor_GAUGE_DISTRIBUTION
+	case otlpmetrics.MetricDescriptor_COUNTER_INT64:
+		return ocmetrics.MetricDescriptor_CUMULATIVE_INT64
+	case otlpmetrics.MetricDescriptor_COUNTER_DOUBLE:
+		return ocmetrics.MetricDescriptor_CUMULATIVE_DOUBLE
+	case otlpmetrics.MetricDescriptor_CUMULATIVE_HISTOGRAM:
+		return ocmetrics.MetricDescriptor_CUMULATIVE_DISTRIBUTION
+	case otlpmetrics.MetricDescriptor_SUMMARY:
+		return ocmetrics.MetricDescriptor_SUMMARY
+	default:
+		return invalidMetricDescriptorType
+	}
+}
+
+func dataPointsToTimeseries(metric *otlpmetrics.Metric) []*ocmetrics.TimeSeries {
+	length := len(metric.Int64Datapoints) + len(metric.DoubleDatapoints) + len(metric.HistogramDatapoints) +
+		len(metric.SummaryDatapoints)
```

Comment:
According to protocol spec only one of these may be set https://github.com/open-telemetry/opentelemetry-proto/blob/c20698d5bb483cf05de1a7c0e134b7c57e359674/opentelemetry/proto/metrics/v1/metrics.proto#L90
However, I think I like the fact that we try to preserve the data even if it does not strictly confirm to the spec. This is the right approach for Collector.